core/src/main/java/com/graphhopper/routing/weighting/TurnWeighting.java

```diff
@@ -103,19 +100,19 @@ public long calcMillis(EdgeIteratorState edgeState, boolean reverse, int prevOrN
                 ? calcTurnWeight(origEdgeId, edgeState.getBaseNode(), prevOrNextEdgeId)
                 : calcTurnWeight(prevOrNextEdgeId, edgeState.getBaseNode(), origEdgeId));
 
-        if (turnCostsInSeconds == 0 && origEdgeId == prevOrNextEdgeId)
-            return millis + (long) defaultUTurnCost * 1000;
-
         return millis + 1000 * turnCostsInSeconds;
     }
 
     /**
      * This method calculates the turn weight separately.
-     * Be aware that it always returns 0 turn weight unless a turn weight was explicitly set. This means it does
-     * NOT include default u-turn costs and calling methods have to check for u-turns.
-     * todo: this should be cleaned up in #1520.
      */
     public double calcTurnWeight(int edgeFrom, int nodeVia, int edgeTo) {
+        if (edgeFrom == EdgeIterator.NO_EDGE || edgeTo == EdgeIterator.NO_EDGE) {
+            return 0;
+        }
+        if (turnCostExt.isUTurn(edgeFrom, edgeTo)) {
```

Comment:
If there is an explicit turn cost stored in the data - then it will be overwritten from this. Should we document this?

I guess this problem was already the case before this PR but the intention was that it should be picked from the data IMO (at least the name setDefaultUTurnCost implies this ;))

Reply:
Yes this has always been like this and this is why I changed the name from `defaultUTurnCost` to `uTurnCost`. Maybe we should check if this is really important for performance. It saves us from looking up the turn cost value from turn cost extension when `edgeFrom==edgeTo`, but this is only a fraction of all turns (we still look up the value for all pairs of edges where `edgeFrom!=edgeTo`. If we want to do something like default left-turn costs or similar we would not be able to do it the same way, because in this case it would be required to check, e.g. for a turn restriction on a left turn.

I added a comment about this in the TurnWeighting constructor.